dataset/test/data_array_test.cpp

```diff
@@ -91,12 +93,70 @@ TEST_F(DataArrayTest, shadow_attr) {
+                  const CopyPolicy copy_policy, const bool expect_copy) {
+  const auto converted = astype(original, target_dtype, copy_policy);
+
+  EXPECT_EQ(converted.data(), astype(original.data(), target_dtype));
+  EXPECT_EQ(converted.masks(), original.masks());
+
+  EXPECT_TRUE(converted.coords()[Dim::X].is_same(original.coords()[Dim::X]));
+  if (expect_copy) {
+    EXPECT_FALSE(converted.data().is_same(original.data()));
+    EXPECT_FALSE(converted.masks()["m"].is_same(original.masks()["m"]));
+  } else {
+    EXPECT_TRUE(converted.data().is_same(original.data()));
+    EXPECT_TRUE(converted.masks()["m"].is_same(original.masks()["m"]));
+  }
+}
+} // namespace
+
 TEST_F(DataArrayTest, astype) {
   DataArray a(
       makeVariable<int>(Dims{Dim::X}, Shape{3}, Values{1, 2, 3}),
-      {{Dim::X, makeVariable<int>(Dims{Dim::X}, Shape{3}, Values{4, 5, 6})}});
-  const auto x = astype(a, dtype<double>);
-  EXPECT_EQ(x.data(), astype(a.data(), dtype<double>));
+      {{Dim::X, makeVariable<int>(Dims{Dim::X}, Shape{3}, Values{4, 5, 6})}},
+      {{"m", makeVariable<bool>(Dims{Dim::X}, Shape{3},
+                                Values{false, true, true})}});
+  check_astype(a, dtype<double>, CopyPolicy::TryAvoid, true);
+  check_astype(a, dtype<double>, CopyPolicy::Always, true);
+  check_astype(a, dtype<int>, CopyPolicy::TryAvoid, false);
+  check_astype(a, dtype<int>, CopyPolicy::Always, true);
+}
+
+namespace {
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+void check_to_unit(const DataArray &original, const units::Unit target_unit,
+                   const CopyPolicy copy_policy, const bool expect_copy) {
+  const auto converted = to_unit(original, target_unit, copy_policy);
+
+  EXPECT_EQ(converted.data(), to_unit(original.data(), target_unit));
+  EXPECT_EQ(converted.coords()[Dim::X].unit(), units::s);
+  EXPECT_EQ(converted.masks(), original.masks());
+
+  EXPECT_TRUE(converted.coords()[Dim::X].is_same(original.coords()[Dim::X]));
+  if (expect_copy) {
+    EXPECT_FALSE(converted.data().is_same(original.data()));
+    EXPECT_FALSE(converted.masks()["m"].is_same(original.masks()["m"]));
+  } else {
+    EXPECT_TRUE(converted.data().is_same(original.data()));
+    EXPECT_TRUE(converted.masks()["m"].is_same(original.masks()["m"]));
+  }
+}
+} // namespace
+
+TEST_F(DataArrayTest, to_unit) {
+  DataArray a(makeVariable<double>(Dims{Dim::X}, Shape{3},
+                                   Values{1.0, 2.0, 3.0}, units::m),
+              {{Dim::X, makeVariable<int>(Dims{Dim::X}, Shape{3},
+                                          Values{4, 5, 6}, units::s)}},
+              {{"m", makeVariable<bool>(Dims{Dim::X}, Shape{3},
+                                        Values{true, false, true})}});
+  check_to_unit(a, units::mm, CopyPolicy::TryAvoid, true);
+  check_to_unit(a, units::mm, CopyPolicy::Always, true);
+  check_to_unit(a, units::m, CopyPolicy::TryAvoid, false);
+  check_to_unit(a, units::m, CopyPolicy::Always, true);
```

Comment:
Can we move these tests to another file? I think we should have mainly tests of the `class DataArray` here.